opendcs-rest-api/src/main/java/org/opendcs/odcsapi/sec/basicauth/TokenManager.java

```diff
@@ -20,170 +20,266 @@
+			if(!DbInterface.isOracle)
 			{
 				// Now verify that user has appropriate privilege. This only works on Postgress currently:
 				String q = "select pm.roleid, pr.rolname from pg_auth_members pm, pg_roles pr "
-					+ "where pm.member = (select oid from pg_roles where rolname = '" + creds.getUsername() + "') "
-					+ "and pm.roleid = pr.oid";
+						+ "where pm.member = (select oid from pg_roles where rolname = '" + creds.getUsername() + "') "
+						+ "and pm.roleid = pr.oid";
 				ResultSet rs = daoBase.doQuery(q);
 				boolean hasPerm = false;
 				while(rs.next() && !hasPerm)
 				{
 					int roleid = rs.getInt(1);
 					String role = rs.getString(2);
-					Logger.getLogger(ApiConstants.loggerName).info("User '" + creds.getUsername() 
-						+ "' has role " + roleid + "=" + role);
-					if (role.equalsIgnoreCase("OTSDB_ADMIN") || role.equalsIgnoreCase("OTSDB_MGR"))
+					LOGGER.info("User '{}' has role {}={}", creds.getUsername(), roleid, role);
+					if("OTSDB_ADMIN".equalsIgnoreCase(role)
+							|| role.equalsIgnoreCase(DbInterface.getAuthenticatedRole()))
+					{
 						hasPerm = true;
+					}
+				}
+
+				if(!hasPerm)
+				{
+					throw new WebAppException(ErrorCodes.AUTH_FAILED,
+							"User does not have OTSDB_ADMIN or OTSDB_MGR privilege - Not Authorized.");
 				}
-				
-				if (!hasPerm)
-					throw new WebAppException(ErrorCodes.AUTH_FAILED, 
-						"User does not have OTSDB_ADMIN or OTSDB_MGR privilege - Not Authorized.");
 			}
 		}
-		catch (Exception e)
+		catch(Exception e)
+		{
+			LOGGER.atWarn().setCause(e).log("isUserValid - Authentication failed");
+			throw new WebAppException(ErrorCodes.AUTH_FAILED, "DB connection failed with passed credentials.");
+		}
+	}
+
+	private static String getDatabaseUrl(DbInterface dbi) throws WebAppException
+	{
+		try
 		{
-			Logger.getLogger(ApiConstants.loggerName).warning("isUserValid - Authentication failed: " + e);
-			throw new WebAppException(ErrorCodes.AUTH_FAILED, "DB connection failed with passed credentials."); 
+			Connection poolCon = dbi.getConnection();
+			// The only way to verify that user/pw is valid is to attempt to establish a connection:
+			DatabaseMetaData metaData = poolCon.getMetaData();
+			return metaData.getURL();
 		}
-		finally
+		catch(SQLException e)
 		{
-			if (userCon != null)
-				try { userCon.close(); } catch(Exception ex) {}
+			throw new WebAppException(HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
+					"Failed to obtain database URL.", e);
 		}
+	}
 
-		// Make a UserToken and token string. Place in Hashmap & return string.
-		String tokstr = Long.toHexString(rand.nextLong());
-		UserToken userTok = new UserToken(tokstr, creds.getUsername());
-		Logger.getLogger(ApiConstants.loggerName).fine("Added new token for user '" 
-			+ creds.getUsername() + "'=" + tokstr);
-		
-		return userTok;
+	private static UserToken createToken(Credentials creds) throws WebAppException
+	{
+		try
+		{
+			JWTClaimsSet claimsSet = new JWTClaimsSet.Builder()
```

Comment:
I see why you went with a different API.

FYI I went through the same design choices with CDA. Having the service provide a generated JWT is just not required. JWTs are better suited when an external system is used for authorization. After verifying credentials creating a new HttpSession (STIG, and just good security practice) and setting the principal appropriately further verification is not required. This is why CWMS_AAA ended up the way it did. I spent a lot of time on the JWT path and it was just more trouble than it's worth.

The Session ID cookies us SecureRandom to generate the IDs, the likelyhood of guessing them is extremely low. While we don't need high performance this also adding cognitive overhead that isn't as helpful as it first looks (I too went about this far down this path.)

Keeping the Principal, with name, roles, etc, on the Server Side is also an extra layer of security.

Additionally with the way you generate the signing key, while certainly security, prevent the system from maintaining sessions after a restart. Some users may with to do that as it is required store serialized sessions external to the memory space and machine to facilitate API instance redundancy.


Short version:

- Invalidate the current HttpSession and generate a new one
- Generate the principal, which I would expand to include all appropriate information (see CWMS_AAA)
- Set said principal in the new HttpSession by the appropriate means
- The various filters would then pick up the principal and create the appropriate SecurityContexts where required. (sorry I'm still more familiar with this is javalin the JAX-RS; haven't gotten there with LRGS's HTTP interface yet.)

Basically, the server can trust it's own sessions and it makes it more flexible to the downstream user how session stores are implemented.

Additionally, the JWT from external source, can take advantage of this code flow by just being the way a user logs in and interacts with the server; which is most commonly how one would interact.

For that using the filters
1. Use the same postCredentials concept for login, but you'd verify and parse the JWT
2. Create Principal and Session
3. First thing to check in the filter chain is if there is a principal or a session
  4.  if not parse the JWT each request (e.g. machine computation)
  5.  if so, use it, user is most likely in a browser; or really knows how to use CURL.

Reply:
Yeah this makes sense, and is where I stared to think after I wen this far down. I started with figuring out how to parse out the JWT then though it was easy enough to just use a JWT instead of the SecureRandom long. Then realized I would either need to not set the expiration on the JWT (defeating the purpose the JWT) or update the client to use a refresh token, but then that seemed like silly extra work given the JWT isn't from an external service. In addition to not being able to maintain sessions after restart, this implementation doesn't support scaling, which is the other big benefit of the JWT, which means its just an encrypted version of the UserToken of the previous implementation.

If the HttpSession is sufficient, then I'll dial this way back. Thanks for the feedback!